### ISSUE TITLE
Move requireNamespace() check for cyclocomp_linter() to the factory?

### ISSUE THREAD

**Issue by MichaelChirico** [r-lib/lintr#2777]
Currently, the validity check for `cyclocomp_linter()` happens inside the lint loop:

https://github.com/r-lib/lintr/blob/e673ee34a4f750130b2c603cff6b81d5aa6bd48f/R/cyclocomp_linter.R#L25-L33

I think it makes more sense to do so outside the loop, but simply moving it breaks many things in our test suite when {cyclocomp} is not available.

Here's one test in particular that gives me pause:

https://github.com/r-lib/lintr/blob/e673ee34a4f750130b2c603cff6b81d5aa6bd48f/tests/testthat/test-linter_tags.R#L55-L74

Are we OK with just doing `skip_if_not_installed('cyclocomp')` here? Should we keep the `requireNamespace()` check where it is?

**Comment by AshesITR** [comment]
What about adding cyclocomp as a dev dependency? It's in our lint config anyway, no?

**Comment by MichaelChirico** [comment]
Yes, it's there. The issue is the `R-CMD-check-hard` CI which explicitly runs the suite absent `Suggests` deps.

**Comment by AshesITR** [comment]
Oh, I see. In that case I'm fine with skip_if_not_installed.
It there a way to guard us against the test being skipped everywhere?

**Comment by AshesITR** [comment]
Yes, basically assert 100% test code execution for some environment.

One (probably very bad) idea:

Define

```r
test_canary <- function(x) {
  x
}
```

and abuse code coverage statistics via

```r
skip_if_not_installed("cyclocomp")
expect_equal(test_canary(42L), 42L)
```

**Comment by MichaelChirico** [comment]
> One (probably very bad) idea:

Won't the existing codecov suite be enough? We'll get coverage there past any existing `skip_if_not_installed('cyclocomp')`, right?

**Comment by MichaelChirico** [comment]
Here are the other 3 linters in our suite using `requireNamespace`:

https://github.com/r-lib/lintr/blob/22ee8dba67b40bab5e71ef2e31fed525c68a0369/R/literal_coercion_linter.R#L80
https://github.com/r-lib/lintr/blob/22ee8dba67b40bab5e71ef2e31fed525c68a0369/R/object_overwrite_linter.R#L58-L62
https://github.com/r-lib/lintr/blob/22ee8dba67b40bab5e71ef2e31fed525c68a0369/R/unused_import_linter.R#L98

That makes `cyclocomp_linter()` unique in that the core functionality of the linter requires a specific package.

 - `literal_coercion_linter()` can _optionally_ tailor its lint message if {rlang} is installed
 - `object_overwrite_linter()` can flag overwrites from user-specified packages, which it does in the factory (by default it only checks packages which should _always_ be installed)
 - `unused_import_linter()` behaves worse when used on inputs importing certain packages that are not available
